src/main/scala/shine/DPIA/Phrases/VisitAndRebuild.scala

```diff
@@ -92,14 +111,45 @@ object VisitAndRebuild {
     }
   }
 
-  private def visitPhraseTypeAndRebuild(phraseType: PhraseType, v: Visitor): PhraseType = phraseType match {
-    case ExpType(dt, w)                 => ExpType(v.data(dt), v.access(w))
-    case AccType(dt)                => AccType(v.data(dt))
+  def visitPhraseTypeAndRebuild(phraseType: PhraseType, v: Visitor): PhraseType = phraseType match {
+    case ExpType(dt, w)                 => ExpType(visitDataTypeAndRebuild(dt, v), v.access(w))
+    case AccType(dt)                => AccType(visitDataTypeAndRebuild(dt, v))
     case CommType()                 => CommType()
     case PhrasePairType(t1, t2)           => PhrasePairType(visitPhraseTypeAndRebuild(t1, v), visitPhraseTypeAndRebuild(t2, v))
     case FunType(inT, outT)         => FunType(visitPhraseTypeAndRebuild(inT, v), visitPhraseTypeAndRebuild(outT, v))
     case PassiveFunType(inT, outT)  => PassiveFunType(visitPhraseTypeAndRebuild(inT, v), visitPhraseTypeAndRebuild(outT, v))
-    case dft @ DepFunType(x, t)           => DepFunType(x, visitPhraseTypeAndRebuild(t, v))(dft.kn)
+    case DepFunType(x, t)           => x match {
+      case n: NatIdentifier =>
+        DepFunType[NatKind, PhraseType](
+          NatIdentifier(v.nat(n).asInstanceOf[arithexpr.arithmetic.NamedVar].name), visitPhraseTypeAndRebuild(t, v))
+      case dt: DataTypeIdentifier =>
+        DepFunType[DataKind, PhraseType](
+          v.data(dt).asInstanceOf[DataTypeIdentifier], visitPhraseTypeAndRebuild(t, v))
+      case ad: AddressSpaceIdentifier =>
+        DepFunType[AddressSpaceKind, PhraseType](
+          v.addressSpace(ad).asInstanceOf[AddressSpaceIdentifier], visitPhraseTypeAndRebuild(t, v))
+      case ac: AccessTypeIdentifier =>
+        DepFunType[AccessKind, PhraseType](
+          v.access(ac).asInstanceOf[AccessTypeIdentifier], visitPhraseTypeAndRebuild(t, v))
+      case n2n: NatToNatIdentifier =>
+        DepFunType[NatToNatKind, PhraseType](
+          v.natToNat(n2n).asInstanceOf[NatToNatIdentifier], visitPhraseTypeAndRebuild(t, v))
+      case n2d: NatToDataIdentifier =>
+        DepFunType[NatToDataKind, PhraseType](
+          v.natToData(n2d).asInstanceOf[NatToDataIdentifier], visitPhraseTypeAndRebuild(t, v))
+    }
+  }
+
+  def visitDataTypeAndRebuild(dataType: DataType, v: Visitor): DataType = dataType match {
+    case i: IndexType => IndexType(v.nat(i.size))
+    case a: ArrayType =>
+      ArrayType(v.nat(a.size), visitDataTypeAndRebuild(a.elemType, v))
+    case vec: VectorType =>
+      VectorType(v.nat(vec.size), v.data(vec.elemType))
+    case r: PairType =>
+      PairType(visitDataTypeAndRebuild(r.fst, v),
+        visitDataTypeAndRebuild(r.snd, v))
+    case _ => v.data(dataType)
```

Comment:
What if `v.data` already deals with the previous cases? then this would ignore it.

Reply:
Updated `visitDataTypeAndRebuild` to use the result from `v.data`, also made it `private` to avoid infinite recursive calls caused by using `visitDataTypeAndRebuild` in `data`.